tiledb/cloud/dag/dag.py

```diff
@@ -169,49 +171,193 @@ def depends_on(self, node: "Node"):
         :param node: node to mark as a dependency of this node
         :return:
         """
-        self.parents[node.id] = node
-        node.children[self.id] = self
+        with self._lifecycle_condition:
+            if self._status is not Status.NOT_STARTED:
+                raise RuntimeError("Cannot add dependency to an already-started node.")
+            self.parents[node.id] = node
+            node.children[self.id] = self
 
-        if self.dag is None and node.dag is not None:
-            self.dag = node.dag
+            if self.dag is None and node.dag is not None:
+                self.dag = node.dag
 
-    def cancel(self) -> bool:
-        return self._future.cancel()
+    # FutureLike methods.
 
-    def finished(self) -> bool:
+    def cancel(self) -> bool:
+        with self._lifecycle_condition:
+            if self._status is not Status.NOT_STARTED:
+                return False
+            self._update_status(Status.CANCELLED)
+            self._lifecycle_exception = futures.CancelledError()
+            cbs = self._callbacks()
+        futures.execute_callbacks(self, cbs)
+        return True
+
+    def done(self) -> bool:
+        with self._lifecycle_condition:
+            return self._done()
+
+    def cancelled(self) -> bool:
+        with self._lifecycle_condition:
+            return self._status is Status.CANCELLED
+
+    def running(self) -> bool:
+        with self._lifecycle_condition:
+            return self._status is Status.RUNNING
+
+    def result(self, timeout: Optional[float] = None) -> _T:
+        with self._lifecycle_condition:
+            self._wait(timeout)
+            to_raise = self._lifecycle_exception or self._exception
+            if to_raise:
+                raise to_raise
+            result = self._result
+        assert result
+        return result.get()
+
+    def exception(self, timeout: Optional[float] = None) -> Optional[BaseException]:
+        with self._lifecycle_condition:
+            self._wait(timeout)
+            if self._lifecycle_exception:
+                raise self._lifecycle_exception
+            return self._exception
+
+    def add_done_callback(self, fn: Callable[["Node[_T]"], None]) -> None:
+        with self._lifecycle_condition:
+            self._cb_list.append(fn)
+            if not self._done():
+                return
+        fn(self)
```

Comment:
The respective `Dag` method guard this in try/except (+ also checks that `fn` is indeed callable); is there a reason they're not identical?

Reply:
Added a try-block here. I omitted the check found in `add_done_callback` because it's kind of superfluous, given that we have the type annotations to guide the user.